pkg/shell/base_index.js

```diff
@@ -496,7 +498,84 @@ function Index() {
     if (typeof self.navigate !== "function")
         throw Error("Index requires a prototype with a navigate function");
 
-    self.frames = new Frames(self);
+    /* Session timing out after inactivity */
+    let session_timer = null;
+    let session_final_timer = null;
+    let session_timeout = 0;
+    let final_countdown = 30000; // last 30 seconds
+    let title = "";
+
+    function startTimer() {
+        if (session_timeout > 0)
+            session_timer = window.setTimeout(sessionTimeout, session_timeout - final_countdown);
+    }
+
+    function sessionTimeout() {
+        /* When non-password login is in use, we don't want to logout used, as the user can just log
+         * back in. This can be recognized by checking `privileges` but this is set up
+         * asynchronously and it may happen that it is not yet set up in StartTimer().
+         * We can just check this out on first sessionTimeout() and if non-pwd login is used then
+         * just remove the timer and do not bother with it anymore.
+         */
+        if (self.privileges.is_locked() == true) {
+            window.clearTimeout(session_timer);
+            return;
+        }
+
+        title = document.title;
+        sessionFinalTimeout();
+        $("#session-timeout-dialog").modal("show");
+
+        document.getElementById("keep-session-alive").addEventListener("click", e => {
+            window.clearTimeout(session_final_timer);
+            session_final_timer = null;
+            document.title = title;
+            resetTimer();
+            $("#session-timeout-dialog").modal("hide");
+            final_countdown = 30000;
+        });
+    }
+
+    function updateFinalCountdown() {
+        const remaining_secs = Math.floor(final_countdown / 1000);
+        const timeout_text = cockpit.format(_("You will be logged out in $0 seconds."), remaining_secs);
+        document.getElementById("timeout-message").innerHTML = timeout_text;
+        document.title = "(" + remaining_secs + ")" + title;
+    }
+
+    function sessionFinalTimeout() {
+        final_countdown -= 1000;
+        if (final_countdown > 0) {
+            updateFinalCountdown();
+            session_final_timer = window.setTimeout(sessionFinalTimeout, 1000);
+        } else {
+            cockpit.logout(true, _("You have been logged out due to inactivity."));
+        }
+    }
+
+    function resetTimer() {
+        if (!session_final_timer) {
+            window.clearTimeout(session_timer);
+            startTimer();
+        }
+    }
+
+    function setupTimers(document) {
+        document.addEventListener("mousemove", resetTimer, false);
```

Comment:
Not sure if merely moving the mouse around should count as "doing" stuff, but that's not that important.

Reply:
I would say that we are checking if someone "is around" and not if someone is "doing something". And mouse moving is great way how to see if someone is around.